amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +153,224 @@ public void stop() {
+                                .apply(streamConfigEntry.getValue());
+                        if (!shardSyncTaskManager.castShardSyncTask()) {
+                            log.warn(
+                                    "Failed to submit shard sync task for stream {}. This could be due to the previous shard sync task not finished.",
+                                    shardSyncTaskManager.shardDetector().streamIdentifier().streamName());
+                        }
+                    } else {
+                        log.info("Skipping shard sync for {} as either hash ranges are complete in the lease table or leases hole confidence is not achieved.", streamConfigEntry.getKey());
+                    }
                 }
+            } catch (Exception e) {
+                log.error("Caught exception while running periodic shard syncer.", e);
             }
         } else {
             log.debug(String.format("WorkerId %s is not a leader, not running the shard sync task", workerId));
         }
     }
 
+    private Map<StreamIdentifier, List<Lease>> getStreamToLeasesMap(
+            final Set<StreamIdentifier> streamIdentifiersToFilter)
+            throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        final List<Lease> leases = leaseRefresher.listLeases();
+        if (!isMultiStreamingMode) {
+            Validate.isTrue(streamIdentifiersToFilter.size() == 1);
+            return Collections.singletonMap(streamIdentifiersToFilter.iterator().next(), leases);
+        } else {
+            final Map<StreamIdentifier, List<Lease>> streamToLeasesMap = new HashMap<>();
+            for (Lease lease : leases) {
+                StreamIdentifier streamIdentifier = StreamIdentifier
+                        .multiStreamInstance(((MultiStreamLease) lease).streamIdentifier());
+                if (streamIdentifiersToFilter.contains(streamIdentifier)) {
+                    streamToLeasesMap.computeIfAbsent(streamIdentifier, s -> new ArrayList<>()).add(lease);
+                }
+            }
+            return streamToLeasesMap;
+        }
+    }
+
+    @VisibleForTesting
+    boolean shouldDoShardSync(StreamIdentifier streamIdentifier, List<Lease> leases) {
+        if (CollectionUtils.isNullOrEmpty(leases)) {
+            // If the leases is null or empty then we need to do shard sync
+            log.info("No leases found for {}. Will be triggering shard sync", streamIdentifier);
+            return true;
+        }
+        // Check if there are any holes in the leases and return the first hole if present.
+        Optional<HashRangeHole> hashRangeHoleOpt = hasHoleInLeases(streamIdentifier, leases);
+        if (hashRangeHoleOpt.isPresent()) {
+            // If hole is present, check if the hole is detected consecutively in previous occurrences.
+            // If hole is determined with high confidence return true; return false otherwise
```

Comment:
Can we add more detail to the comment about why a few holes are okay and we need to hit the threshold to have confidence that we need to do a shard sync?